### ISSUE TITLE
Playback delayed (was Queued Songs Skip, Stutter, or Stop)

### ISSUE THREAD

**Issue by leefc01** [music-assistant/hass-music-assistant#3023]
### What version of Music Assistant has the issue?

2.2.7

### What version of the Home Assistant Integration have you got installed?

2024.9.1

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

When playing any queued music (i.e., album, playlist), songs are skipped.  Playback can also be stuttered or stopped.
This occurs with YouTube Music and Sonos.

### How to reproduce

1.  Open MusicAssistant
2. Search for any album. See Album Details.  Example:  Miles Davis - Kind of Blue., 
3. Select from drop down, Play Now (clear queue).
4. Playback will start.  Album tracks will be queued.
5. Consistently, upon completion of Track 1, playback will skip to Track 5.

Other albums and playlists will skip to different track numbers. 

### Music Providers

YouTube Music
I don't have other subscriptions at this time.

### Player Providers

Sonos
Don't have other target players at the moment.  Seem to be problems adding Squeezebox Boom.

### Full log output

[INFO log for Miles Davis - Kind of Blue](https://github.com/user-attachments/files/17285346/music-assistant.info.log)

Supplemental
- [DEBUG log for Miles Davis - Kind of Blue](https://github.com/user-attachments/files/17285345/music-assistant.debug.log)

- [INFO](https://github.com/user-attachments/files/17285343/music-assistant.yo.debug.log) and [DEBUG](https://github.com/user-attachments/files/17285344/music-assistant.yo.info.log) logs for Yo Yo Ma - Six Evolutions 

### Additional information

1. What music provider is in use when the problem is observed? What other music providers have you tried (if you have no others then make that clear)
- YouTube Music.  No other subscriptions.

2. What player provider is in use when the problem is observed? What other player providers have you tried (if you have no others then make that clear) 
- Sonos - issue observed
- Squeezebox Boom - pending resolving adding to system.  Currently erroring while adding.

3. Are the players grouped?
- No, but grouped players will exhibit the same problems

5. How is playback being instigated (e.g. automation or via the UI)
- UI, both web and mobile

6. If you aren't using HAOS then exactly how you have installed MA
- NA, using HAOS

7. What is working (e.g. player works with music provider X but not Y)
- Playing a single song works via YouTube Music
- Search, forward, reverse, play, stop, pause all function as expected

8. What you have tried from this page that hasn't **helped**
- Tried troubleshooting FAQs, started MA, HA, and host independently, in that order.
- Deleted core cache
- Verified no physical device settings impacting play
- Confirmed Shuffle is disabled
- Confirmed correct IPs
- Checked Discord.  At least 1 other user reported similar issues with YouTubeMusic and Sonos

Log Observation:  Network connection was reset by peer and there were premature disconnections.

Other Notes:
- Running HAOS on a Mini PC / Proxmox Ubuntu / VM
- No VLANs

### What version of Home Assistant Core are your running

2024.10.1

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Generic x86-64 (e.g. Intel NUC)

**Comment by OzGav** [comment]
Please try Snapcast and stream to a browser so we can see if this is a sonos problem

**Comment by teijosantala** [comment]
For me the skipping occurs with Sonos and remote share.
Using Home Assistant MediaPlayers and enabling the queue flow mode makes the stream play normally without skipping but the ui/music assistant will still think that it skipped songs rendering the progress bar and playlist unusable.

**Comment by leefc01** [comment]
> Please try Snapcast and stream to a browser so we can see if this is a sonos problem

Plabyack via Snapcast works as expected.  No disruptions.
[MA Log Using Snapcast INFO](https://github.com/user-attachments/files/17334522/music-assistant.snap.kind.info.log)

Retested via Sonos are playback problems persist.  Logs show "premature client disconnection."
[MA Log Using Sonos DEBUG](https://github.com/user-attachments/files/17334529/music-assistant.sonos.debug.log)

**Comment by OzGav** [comment]
OK it is likely the issue here is that you are still running the stable server whereas significant improvements have been made in beta. Specifically there are now two providers, one for Series 1 speakers and one for Series 2. You will either need to wait for a future stable release or try and the beta and let us know if that resolves things. You can run both stable and beta on the same host just not at the same time.

**Comment by leefc01** [comment]
> OK it is likely the issue here is that you are still running the stable server whereas significant improvements have been made in beta. Specifically there are now two providers, one for Series 1 speakers and one for Series 2. You will either need to wait for a future stable release or try the beta and let us know if that resolves things. You can run both stable and beta on the same host just not at the same time.

Thanks for the update.  Testing so far shows the playback issue is resolved in Beta.   And now I know how to get the Beta version are digging a bit.

Separately, perhaps just in beta is that it takes nearly 1 minute before playback starts.   I'll hold off on further analysis since it's still in beta.

**Comment by OzGav** [comment]
No one else is reporting the slow stream start so please look further into that. Advise exactly which provider you are using and what device as well.

**Comment by HemiBob** [comment]
I have been experiencing the exact same behaviour so I'm very pleased that a fix is on the way.
For info, the music source is Synology NAS and the player is Sonos S2.

**Comment by WhiteJamal** [comment]
Experiencing issue using HASS with VLC addon. Speaker is connected via AUX cable to computer. When I add music to queue from Apple Music or soundcloud the song either restarts instead of continuing to play or skips the queue. I either click 'Play Next' Or 'Add to queue' and still get this bug.

**Comment by leefc01** [comment]
> No one else is reporting the slow stream start so please look further into that. Advise exactly which provider you are using and what device as well.

As a follow up, I'm seeing delayed responses on the MA Beta 2.3.0b29.

I initiate playback with the same process in the original bug:
1. Search for album from the Web UI
2. From the dropdown, select Play Now (Clear Queue)
3. I used the same album to test - Miles Daves:  Kind of Blue
4. Playback start is delayed ~20-30 seconds, but all tracks play.  Only the first track has delayed playback.

Provider:  YouTubeMusic
Player:  Sonos

Let me know if you'd rather I write this up as a separate issue.

Logs
INFO logs didn't appear to be useful, so I included DEBUG versions.  I noticed that the logs show a ~20-30s delay between receiving a play command and initiating the stream.  

I also checked the CPU/RAM at this time.  CPU barely moved from <1% usage and RAM never went above 7%.  

I did see a few "premature client disconnection" in the logs, but it didn't impact playback.
 
[Miles Davis - INFO](https://github.com/user-attachments/files/17438894/music-assistant.miles.info.log)
[Miles Davis - DEBUG(https://github.com/user-attachments/files/17438898/music-assistant.delay.miles.debug.log)
[The Linda Lindas - DEBUG](https://github.com/user-attachments/files/17438896/music-assistant.delay.linda.debug.log)
[Assorted test cases - DEBUG](https://github.com/user-attachments/files/17438895/music-assistant.delay.debug.log)

Perhaps coincidentally, on one of the tests, I see that YouTube Music timed out.  MA recovered from this, but adding as supplemental.
[YouTubeMusic Timeout - DEBUG](https://github.com/user-attachments/files/17438899/music-assistant.YTMtimeout.debug.log)

**Comment by OzGav** [comment]
@leefc01 OK please try another player provider and see whether the delayed response is the same there. Is your original post resolved in the beta?

**Comment by OzGav** [comment]
@WhiteJamal This issue relates to Sonos so you need to bring up your problem elsewhere. Did you look at this https://github.com/music-assistant/hass-music-assistant/issues/2813

**Comment by leefc01** [comment]
> 

Original issue is resolved in current beta 2.3.0b30.

Also retested playback using Snapcast and Sonos.  No significant delay noted.
[Log for referencer - DEBUG](https://github.com/user-attachments/files/17445839/music-assistant.2.3.0b30.delay.log)

**Comment by marcelveldt** [comment]
@leefc01 so if I understand you correctly, the issue has been resolved with 2.3.0 beta 30 and later ?

**Comment by leefc01** [comment]
> @leefc01 so if I understand you correctly, the issue has been resolved with 2.3.0 beta 30 and later ?

The original playback issue is resolved in 2.3.0 beta 30.  This is where songs are skipped, stuttered, or stopped.  

However, in testing the beta, I observed a ~20-30s delay when initiating playback.  This is using YouTube Music / Sonos.  The problem is not observed with Snapcast / Browser.

**Comment by marcelveldt** [comment]
Is that with Sonos S1 provider in MA or S2 ?

**Comment by leefc01** [comment]
@marcelveldt 
> Is that with Sonos S1 provider in MA or S2 ?

S2

**Comment by leefc01** [comment]
Conducted 2 other tests.

1. YouTube Music / Slimproto via MA - no delay issues for initiating playback.  

2. YouTube Music / Sonos via Home Assistant native - no delay issues for initiating playback.  

I noticed in MA that there seemed to be a delay in initiating playback as well as acting on any other player commands (i.e., pause, stop, volume up / down).

**Comment by marcelveldt** [comment]
> YouTube Music / Sonos via Home Assistant native - no delay issues for initiating playback.

What do you mean by this ? Using the HA media player for Sonos into MA ?

> I noticed in MA that there seemed to be a delay in initiating playback as well as acting on any other player commands (i.e., pause, stop, volume up / down).

I have no idea whats going on but both me and @MarvinSchenkel are not able to reproduce these issues on Sonos speakers with the S2 provider in the MA beta (33). Is there some logging in the server when you try sending commands ?

**Comment by leefc01** [comment]
> What do you mean by this ? Using the HA media player for Sonos into MA ?
Yes 
 
> I have no idea whats going on but both me and @MarvinSchenkel are not able to reproduce these issues on Sonos speakers with the S2 provider in the MA beta (33). Is there some logging in the server when you try sending commands ?

There's not a lot going on in the logs.  I set both MA and Sonos to DEBUG.

```
2024-10-22 15:27:23.604 DEBUG (MainThread) [music_assistant.players] Handling command play_media for player Portable
2024-10-22 15:27:23.604 DEBUG (MainThread) [music_assistant.players] Handling command cmd_power for player Portable
2024-10-22 15:27:43.928 DEBUG (MainThread) [music_assistant.streams] Start serving audio stream for QueueItem Miles Davis - So What (feat. John Coltrane, Cannonball Adderley & Bill Evans) (library://track/315) to Portable
2024-10-22 15:27:43.939 DEBUG (MainThread) [music_assistant.audio.media_stream] start media stream for: ytmusic--ysxLwfnM://track/9rD3zdscDRQ
2
```

I also took a screen recording and attached the log to demonstrate what I'm seeing in Beta 34.

[Screen Recording](https://drive.google.com/file/d/1XhQgsBwdciBHq9n92Xz1L-WBd4yXBJxM/view?usp=share_link)
[DEBUG](https://github.com/user-attachments/files/17482788/music-assistant.delay.screenshot.debug.log)

**Comment by marcelveldt** [comment]
You shouldnt use Sonos through HA mediaplayers, that is doomed to fail as a lot of special sonos handling will be missing then.
The native MA player providers for e.g. Sonos are tweaked endlessly to give the best playback experience.

Please give this another test using the native MA sonos provider once 2.3.0 is released tomorrow.
Thanks!

**Comment by leefc01** [comment]
Tested 2.3.0 and the results are the same as with the beta.
1. Original issue of Queued songs skipping, stuttering, or stopping has been resolved.
2. Observed a delayed reaction of Sonos speakers handling commands (e.g., Initiate album playback, pause, change volume).  The logs only show a gap between actions.  ~20 seconds before starting an album, ~ 10 seconds for pausing.

Let me know if you want to close the original issue.

I'll keep experimenting.  Would be interested if any other users experience this.
```
2024-10-25 08:10:18.327 DEBUG (MainThread) [music_assistant.players] Handling command play_media for player Portable
2024-10-25 08:10:38.457 DEBUG (MainThread) [music_assistant.streams] Start serving audio stream for QueueItem Miles Davis - So What (feat. John Coltrane, Cannonball Adderley & Bill Evans) (library://track/315) to Portable
2024-10-25 08:10:38.465 DEBUG (MainThread) [music_assistant.audio.media_stream] start media stream for: ytmusic--fHWEem35://track/9rD3zdscDRQ
2024-10-25 08:10:48.736 DEBUG (MainThread) [music_assistant.audio] Getting streamdetails for library://track/316
2024-10-25 08:10:50.127 DEBUG (MainThread) [music_assistant.ytmusic] Found stream_format: 141 - audio only (high) for song F-SuJ3bLkLI
2024-10-25 08:10:50.132 DEBUG (MainThread) [music_assistant.audio] retrieved streamdetails for library://track/316 in 1396 milliseconds
2024-10-25 08:10:50.987 DEBUG (MainThread) [music_assistant.webserver] Handling command players/cmd/play_pause
2024-10-25 08:10:50.988 DEBUG (MainThread) [music_assistant.players] Handling command cmd_pause for player Portable
2024-10-25 08:11:01.202 DEBUG (MainThread) [aiohttp.server] Ignored premature client disconnection 2
2024-10-25 08:11:01.202 DEBUG (MainThread) [aiohttp.server] Ignored premature client disconnection.
2024-10-25 08:11:01.220 DEBUG (MainThread) [music_assistant.audio.media_stream] stream aborted (with code 255) for ytmusic--fHWEem35://track/9rD3zdscDRQ - seconds streamed: 37.0
```

**Comment by OzGav** [comment]
@leefc01 confirm that your testing above was using the native MA Sonos player provider. If so, please advise which one - Sonos or Sonos S1

**Comment by leefc01** [comment]
> @leefc01 confirm that your testing above was using the native MA Sonos player provider. If so, please advise which one - Sonos or Sonos S1

Yes.  Using native MA Sonos player provider.  Sonos, not S1.

I'll continue testing the release as well as the beta and provide logs tomorrow for reference.

**Comment by leefc01** [comment]
Same results from a new HAOS VM with MA as the only add on.  

1. Created a new, separate HAOS VM using Proxmox.  
2. Added MA as the only add-on.  
3. Added Sonos as the only Player
4. Added YouTube Music as the only Provider
5. Initiated playback and experienced the same delayed playback.

**Comment by marcelveldt** [comment]
Multiple people have tried to reproduce this issue without success. 
Could it be that something is throttling network traffic speed between your HA/proxmox host and the speakers ?
Did you also test it from e.g. a spare pi or other dedicated hardware to rule out some proxmox issue ?

Sonos speakers require a certain amount of buffer fullness before they start playing. The symptoms you are describing match that of the buffer being filled too slowly and then the playback also starts later.

**Comment by leefc01** [comment]
Thanks for trying to reproduce.

There should be no throttling other than what the hardware is capable of. No special rules I've setup on the network.  MA server is connected via Ethernet. 

I do have a spare Pi 4b.  I'll setup HA / MA and retest.  This will take a couple days as I'm out of town.  

I did do a test with MA / Slimproto and there were no issues. 

Will update after testing.

**Comment by selmaohneh** [comment]
I am experiencing similar delays. For me it is on skipping to the next track. It takes up to 30s-60s before MA notices the `media_player: next` service call and the next track actually starts. 

![image](https://github.com/user-attachments/assets/90acb0f0-e19e-484d-b452-44c79e1a8236)

I use a Sonos 1 via the native MA provider. Spotify is my music provider.

**Comment by leefc01** [comment]
Good news, bad news.

Good - I no longer have delayed responses to commands (Start, Pause, Next, Previous, Volume Up / Down).
Bad - I'm not sure why.

Longer notes in case this helps anyone.

I initially tested the following cases and experienced delayed responses from Sonos via MA.
- Proxmox / HAOS / MA / YouTube Music / Sonos
- Proxmox w/ new VM and only HAOS / MA / YouTube Music / Sonos
- RPi3 / HAOS / MA / YouTube Music / Sonos
- Proxmox / HAOS / MA / Apple Music / Sonos
- Proxmox / HAOS / MA / Plex / Sonos

Use cases with no delay
 - Proxmox / HAOS / MA / YouTube Music / Snapcast / Web
 - Proxmox / HAOS / MA / YouTube Music / Slimproto / Squeezebox Boom and Squeezebox Radio
 - Proxmox / HAOS / MA / Apple Music / Slimproto / Squeezebox Boom and Squeezebox Radio

I suspected MA might be making calls to Sonos.com before streaming.  To test, I decided to install RPi.

While installing and configuring RPi amid doing other things today, I noticed that there were updated versions of HA Core and Supervisor.  I installed them.

All ready for my test using track DNS calls with RPi, my first MA test showed no delay.  

I'll just take that it works currently.

**Comment by OzGav** [comment]
Thanks so much for the extensive testing. Will move this to DONE but keep it open for a while until things looks confirmed fixed.

**Comment by OzGav** [comment]
@leefc01 can this be closed?

**Comment by leefc01** [comment]
Issue no longer present.  Thanks for all the help.
